csharp/ql/src/semmle/code/csharp/dataflow/internal/rangeanalysis/SignAnalysisCommon.qll

```diff
@@ -242,30 +257,105 @@ Sign ssaDefSign(SsaVariable v) {
   )
 }
 
+/** Returns the sign of explicit SSA definition `v`. */
+private Sign explicitSsaDefSign(SsaVariable v) {
+  exists(VariableUpdate def | def = getExplicitSsaAssignment(v) |
+    result = exprSign(getExprFromSsaAssignment(def))
+    or
+    anySign(result) and explicitSsaDefWithAnySign(def)
+    or
+    result = exprSign(getIncrementOperand(def)).inc()
+    or
+    result = exprSign(getDecrementOperand(def)).dec()
+  )
+}
+
+/** Returns the sign of implicit SSA definition `v`. */
+private Sign implicitSsaDefSign(SsaVariable v) {
+  result = fieldSign(getImplicitSsaDeclaration(v))
+  or
+  anySign(result) and nonFieldImplicitSsaDefinition(v)
+}
+
+/** Gets a possible sign for `f`. */
+private Sign fieldSign(Field f) {
+  if not fieldWithUnknownSign(f)
+  then
+    result = exprSign(getAssignedValueToField(f))
+    or
+    fieldIncrementOperationOperand(f) and result = fieldSign(f).inc()
+    or
+    fieldDecrementOperationOperand(f) and result = fieldSign(f).dec()
+    or
+    result = specificFieldSign(f)
+  else anySign(result)
+}
+
 /** Gets a possible sign for `e`. */
 cached
 Sign exprSign(Expr e) {
   result = certainExprSign(e)
   or
   not exists(certainExprSign(e)) and
   (
-    unknownSign(e)
+    anySign(result) and unknownSign(e)
     or
-    exists(SsaVariable v | getARead(v) = e | result = ssaVariableSign(v, e))
+    exists(SsaVariable v | getARead(v) = e |
+      result = ssaSign(v, any(SsaReadPositionBlock bb | getAnExpression(bb) = e))
+      or
+      not exists(SsaReadPositionBlock bb | getAnExpression(bb) = e) and
+      result = ssaDefSign(v)
+    )
     or
-    e =
-      any(VarAccess access |
-        not exists(SsaVariable v | getARead(v) = access) and
-        (
-          result = fieldSign(getField(access.(FieldAccess))) or
-          not access instanceof FieldAccess
-        )
+    exists(VarAccess access | access = e |
+      not exists(SsaVariable v | getARead(v) = access) and
+      (
+        result = fieldSign(getField(access.(FieldAccess)))
+        or
+        anySign(result) and not access instanceof FieldAccess
       )
+    )
     or
     result = specificSubExprSign(e)
   )
 }
 
+/** Gets a possible sign for `e` from the signs of its child nodes. */
+private Sign specificSubExprSign(Expr e) {
+  result = exprSign(getASubExpr(e))
+  or
+  e =
+    any(DivExpr div |
```

Comment:
@hvitved seem to like `any`. I have no preference. But in this case, maybe the `exists` is a better choice, because the succeeding `or` subclause is using that too.